Feature/user group ui

esotrope opened the pull request:
Not satisfied with the UX but it gets the job done. However, I figured this would be a good starting point for discussion.

TODOs: integrate advanced filters component as appears on the exam result page

![image](https://user-images.githubusercontent.com/23462925/40629449-06122b18-6280-11e8-8db8-61981b2aed87.png)

![image](https://user-images.githubusercontent.com/23462925/40629455-1138cae2-6280-11e8-8591-ec4163379834.png)

![image](https://user-images.githubusercontent.com/23462925/40629463-24b4a6d6-6280-11e8-8b3a-316e9fe58b62.png)

esotrope commented on reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java:
```diff
@@ -100,7 +100,7 @@ public Integer getStudentCount() {
      * @return <code>true</code> if the group was created by a user instead of an admin for the user
      */
-    private boolean isUserCreated() {
+    public boolean isUserCreated() {
```
bugfix

esotrope commented on webapp/src/main/webapp/src/app/student/search/student-search-form.component.html:
```diff
@@ -0,0 +1,59 @@
+
+    <div class="flex-child grow">
+      <div class="form-group">
+        <label>Student Name / SSID</label>
```
plan on cleaning up the hardcoded text once the layout etc is more final

esotrope commented on webapp/src/main/webapp/src/app/user-group/user-group.component.ts:
```diff
@@ -0,0 +1,232 @@
+          && (
+            student.ssid.startsWith(this.studentForm.name)
+            || `${(student.lastName + student.firstName).toLowerCase()}`.includes(nameSearch)
+            || `${(student.firstName + student.lastName).toLowerCase()}`.includes(nameSearch)
```
implemented the student search by name/ssid on the client side with a somewhat forgiving matching strategy, not sure what we think about this.

esotrope commented on webapp/src/main/webapp/src/app/user-group/user-group.component.ts:
```diff
@@ -0,0 +1,232 @@
+    return this.processingSubscription != null
+      || this.groupForm == null
+      || !this.groupForm.formGroup.valid
+      || equals(this.originalGroup, this.group);
```
only lets you save if there was a change to the group

esotrope commented on webapp/src/main/webapp/src/app/user-group/user-group.component.ts:
```diff
@@ -0,0 +1,232 @@
+    // TODO only save when there are changes
+    this.processingSubscription = this.service.saveGroup(this.group)
+      .subscribe(() => {
+          this.navigateHome();
```
navigates the user away from the form on save delete and cancel to combat the "apply every change" kind of use

agorina commented on reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java:
```diff
@@ -163,8 +163,8 @@ public Group build() {
-            group.studentIds = studentIds != null ? ImmutableSet.copyOf(studentIds) : ImmutableSet.of();
-            group.students = students != null ? ImmutableSet.copyOf(students) : ImmutableSet.of();
+            group.studentIds = studentIds != null ? ImmutableSet.copyOf(studentIds) : null;
+            group.students = students != null ? ImmutableSet.copyOf(students) : null;
```
Maybe add a helper to `org.opentestsystem.rdw.reporting.common.util.CollectionUtils`?

esotrope commented:
* Redesigned layout
* Added advanced filters
* Fixed form validation

Still todo:
1. Add invalid field text
2. Add delete modal

![image](https://user-images.githubusercontent.com/23462925/40750212-6602264e-641b-11e8-9780-e6aa5d161d8c.png)

![image](https://user-images.githubusercontent.com/23462925/40750241-7eabbd54-641b-11e8-97e9-8f2c8606f85b.png)

![image](https://user-images.githubusercontent.com/23462925/40750257-8dab0b5c-641b-11e8-9a4e-e96925ae7272.png)

![image](https://user-images.githubusercontent.com/23462925/40750283-a83106d4-641b-11e8-9767-ac187dd94379.png)
